### PR TITLE
Cross product of patrick test cases

### PR DESCRIPTION
**Pull request by joba00002**
This will be a useful utility for several new tests that are planned.

Signed-off-by: Jonathan Baumann <joba00002@stud.uni-saarland.de>

### Prerequisites

- [x] I adhere to the coding conventions (described [here](../CONTRIBUTING.md)) in my code.
- [X] I have updated the copyright headers of the files I have modified.
- [X] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [X] I have put signed-off tags in *all* commits.
- [X] I have updated the changelog file [NEWS.md](../NEWS.md) appropriately.
- [X] I have checked whether I need to adjust the showcase file `showcase.R` with respect to my changes.
- [X] The pull request is opened against the branch `dev`.

### CODE REVIEW

File: tests/testing-utils.R [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+## with this program; if not, write to the Free Software Foundation, Inc.,
+## 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+##
+## Copyright 2022 by Jonathan Baumann <joba00002@stud.uni-saarland.de>
+## All Rights Reserved.
+
+
+#' Construct the 'cross product' of two patrick::cases objects.
+#' Each case of the left object is combined with each case of the right,
+#' test names are joined with a comma.
```
Comment (hechtlC) [inline tests/testing-utils.R:20]: Left and right could be confusing as the actual parameters are called `cases.1` and `cases.2`. So please either change the comment to e.g., first and second case or change the parameters to `cases.left` and `cases.right`.

File: tests/testing-utils.R [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        for (id.2 in seq_along(cases.2)){
+            row.2 = cases.2[id.2,]
+
+            test.name.1 = row.1[[".test_name"]]
+            test.name.2 = row.2[[".test_name"]]
+            ## The new test name should reference both previous names.
+            ## Joining with a comma is an obvious option
+            test.name.combined = paste(test.name.1, test.name.2, sep = ", ")
+
+            ## Select everthing from row.1 and everything but '.test_name' from row.2.
```
[the remark this answers is not in the document]
Comment (hechtlC) [reply tests/testing-utils.R:46]: And additionally typo in 'everthing'.

**Review by hechtlC — COMMENTED** [review]
Looks good so far @joba00002. But there are a couple of things that need to be changed as @bockthom already indicated.

**Comment by bockthom** [comment]
> Where exactly should I put such an example? If by function documentation, you mean the documentation comment above it, then I'm not sure if I would put an example there. It would take up a fair amount of space, although it would be easy to find there.

My idea was to put it in the roxygen2 function documentation, that is, directly *before* explaining the parameters via `@param`.
We don't care about how much space it takes. More importantly, such an example would tell the user how this function can be practically used---and the function documentation would be the place where I would look for examples.

If you would think that such an example would be way too long (what I cannot imagine yet), you can also post the example here first before integrating it into the codebase.

File: tests/testing-utils.R [hunk 3]
```diff
@@ -24,6 +24,25 @@ requireNamespace("patrick")
...
+#' joined.cases = cases.cross.product(patrick::cases(
+#'      "sliding window: FALSE" = list(test.sliding.window = FALSE),
+#'      "sliding window: TRUE" = list(test.sliding.window = TRUE)
+#' ), patrick::cases(
+#'          "PaStA: FALSE" = list(test.pasta = FALSE),
+#'          "PaStA: TRUE" = list(test.pasta = TRUE)
+#'      ))
+#'
+#'  yields the following tibble:
+#' .test_name test.sliding.window test.pasta
```
Comment (bockthom) [inline tests/testing-utils.R:37]: Thanks for the example, it looks good! :+1: 

But please increase the readability be adjusting the spacing:
- Add blank line between "Example:" and "joined.cases = ..."
- Add another blank line between "yields the following tibble" and ".test_name ...."
- Adjust the indentation in the example: Is there a reason for the doubled indentation in the second `patrick::cases`? Especially also the position of the `))` at the end looks weird.

Afterwards, as soon as @hechtlC also agrees, we can merge.

Comment (bockthom) [reply tests/testing-utils.R:37]: And could you please add a statement to the `Added` section of the NEWS.md stating that parameterized tests are now possible, for example "Add testing utillity file `tests/testing-utils.R` to make parameterized tests on the cross-product of multiple parameter values possible" or something similar? (This helps tracking the new functionality back, even if it is just testing functionality. When just enhancing, moving, or adding tests, this is not necessary to add, of course.)

---

And just an additional remark (which you can just ignore for this PR, but which might be worth thinking about): This utility function works for two parameters only, if I saw it correctly. What happens if there would be three parameters? Can I just call `cases.cross.product` nested then? For example: 

`cases.cross.product(patrick::cases({...}, cases.cross.product(patrick::cases({...},{...})`

Would this work, or would I need another workaround in this case?

Comment (hechtlC) [reply tests/testing-utils.R:37]: I agree with the formatting issues brought up by @bockthom. Also the Added section in the NEWS.md file.

**Review by hechtlC — APPROVED** [review]
Looks good @joba00002! Thanks for the addition. As soon as the issues brought up by @bockthom are resolved, I feel good about merging this.

File: tests/testing-utils.R (hunk 3 above)
Comment (joba00002) [reply tests/testing-utils.R]: > And just an additional remark (which you can just ignore for this PR, but which might be worth thinking about): This utility function works for two parameters only, if I saw it correctly. What happens if there would be three parameters? Can I just call `cases.cross.product` nested then? For example:
> 
> `cases.cross.product(patrick::cases({...}, cases.cross.product(patrick::cases({...},{...}))`
> 
> Would this work, or would I need another workaround in this case?

Yes, nested calls work.

**Review by bockthom — APPROVED** [review]
Looks good now. Thanks for testing the "nesting" and fixing the iteration bugs. I also played around with these iterating functions and the solution now seems to work :+1:

**Pull request merged**